aggregator/src/aggregator/key_rotator.rs

```diff
@@ -0,0 +1,911 @@
+                            to_be_expired_keypair.id(),
+                            &HpkeKeyState::Expired,
+                        )
+                        .await?;
+                    }
+                }
+            }
+
+            // Promote any pending keypairs that are ready. We don't sweep them in the same loop
+            // as active keypairs, because it's possible for there to be a pending keypair
+            // without an active one.
+            for pending_keypair in pending_keypairs {
+                if Self::duration_since(clock, pending_keypair.updated_at())
+                    > config.pending_duration
+                {
+                    info!(id = ?pending_keypair.id(), "pending key is ready, moving it to active");
+                    tx.set_global_hpke_keypair_state(pending_keypair.id(), &HpkeKeyState::Active)
+                        .await?;
+                }
+            }
+        }
+
+        for expired_keypair in &expired_keypairs {
+            if Self::duration_since(clock, expired_keypair.updated_at()) > config.expired_duration {
+                info!(id = ?expired_keypair.id(), "deleting expired keypair");
+                tx.delete_global_hpke_keypair(expired_keypair.id()).await?;
+            }
+        }
+
+        Ok(())
+    }
+
+    fn duration_since(clock: &C, time: &Time) -> Duration {
+        // Use saturating difference to account for time skew between key rotator runners. Since
+        // key rotators are synchronized by an exclusive lock on the table, it's possible that
+        // time skew between concurrently running replicas result in underflows.
+        clock.now().saturating_difference(time)
+    }
+
+    fn partition_keypairs<'a>(
+        keypairs: &'a [GlobalHpkeKeypair],
+        ciphersuite: &HpkeCiphersuite,
+    ) -> PartitionedKeypairs<'a> {
+        let mut ret = PartitionedKeypairs::default();
+        keypairs
+            .iter()
+            .filter(|keypair| &keypair.ciphersuite() == ciphersuite)
+            .for_each(|keypair| match keypair.state() {
+                HpkeKeyState::Pending => ret.pending_keypairs.push(keypair),
+                HpkeKeyState::Active => ret.active_keypairs.push(keypair),
+                HpkeKeyState::Expired => ret.expired_keypairs.push(keypair),
+            });
+        ret
+    }
+
+    fn available_ids(current_keypairs: &[GlobalHpkeKeypair]) -> impl Iterator<Item = HpkeConfigId> {
+        let ids: Vec<_> = current_keypairs
+            .iter()
+            .map(|keypair| u8::from(*keypair.id()))
+            .collect();
+
+        // Try to cycle through the entire u8 space before going back to zero. This allows us a
+        // bigger window to quicky reject reports from faulty old clients with an outdated HPKE
+        // config error, rather than attempting to decrypt them with the wrong key.
+        let newest_id = current_keypairs
+            .iter()
+            .max_by(|a, b| a.created_at().cmp(b.created_at()))
```

Comment:
Note that treating the highest config ID as the most recent will eventually result in cycling through 253, 254, and 255, instead of 0, 1, and 2 if we picked the lowest unused ID.